ruby/phase-1.md

```diff
@@ -0,0 +1,536 @@
+class Temporal::WorkflowHandle
+  # Return the result of the workflow execution or raise if workflow is still running
+  # - await: true will wait (for up to :timeout seconds) for the workflow result
+  def result(await: false, timeout: nil) -> Any
+
+  # Returns more information about a workflow
+  def describe() -> Temporal::WorkflowExecutionInfo
+
+  # Sends a cancellation request
+  def cancel() -> void
+
+  # Sends a query and awaits for the response
+  def query(name, *args) -> Any
+
+  # Sends a signal
+  def signal(name, *args) -> void
+
+  # Terminates a workflow
+  def terminate(reason, details: nil) -> void
+end
+```
+
+### Alternatives considered
+
+We thought about configuring the `Temporal::Client` with a namespace, similar to the Python SDK.
+It does bring the benefit of avoiding namespace to subsequent calls to the client. However we feel
+like it is an unjustified and artificial limitation that will impact some multi-tenant applications.
+Besides this can always be brought in the shape of a `NamespacedClient` if there's a need for it.
+
+
+## Worker
+
+The worker interface allows SDK users to start processing registered workflow and activities. These
+will then be polled for on the specified namespaces & task queues and executed.
+
+```ruby
+config = Temporal::Configuration.new(url: 'localhost:7233')
+
+worker_1 = Temporal::Worker.new(config, 'my-namespace', 'my-task-queue')
+worker_1.register_workflow(MyWorkflow) # uses 'MyWorkflow' as a name
+worker_1.register_activity(MyActivity) # uses 'MyActivity' as a name
+worker_1.start
+
+worker_2 = Temporal::Worker.new(config, 'my-namespace', 'another-task-queue')
+worker_2.register_workflow(MyWorkflow, name: 'my-workflow') # uses 'my-workflow' as a name
+worker_2.register_activity(MyActivity, name: 'my-activity') # uses 'my-activity' as a name
+worker_2.start
+
+# later
+worker_1.stop
+worker_2.stop
+```
+
+Notes:
+
+- This is not complete example of all configuration options for a worker
+- A worker will use the workflow/activity class name as a name unless it is explicitly provided (
+  `worker_2` demonstrated this approach)
+- Each worker will run in its own thread
+- A blocking `Worker#run` method will be available for convenience of running a single worker in
+  a process until SIGTERM or SIGINT is received
+- In case connection sharing is a concern we can create and memoize a connection on the instance of
+  Temporal::Configuration
+
+And here is the `Worker` interface:
+
+```ruby
+class Worker
+  # Create a new instance of a Worker
+  def self.new(config: Temporal::Configuration, namespace: String, task_queue: String, **options) -> Worker
+
+  # Register a workflow with an optional :name
+  def register_workflow(Class, name: nil) -> void
+
+  # Register an activity with an optional :name
+  def register_activity(Class, name: nil) -> void
+
+  # Start the worker and block until the process exits
+  def run() -> void
+
+  # Start the worker asynchronously
+  def start() -> void
```

Comment:
A fatal worker error would crash the Thread that it runs in, which we can expose via a separate getter. Another feature (postponed till phase 2) is a hook for handling errors, which can help here. Basically most apps would want to hook into that and send the exception data to some 3rd party service (Bugsnag, Airbrake, Sentry, DataDog, etc)

Reply:
I don't know much about the ruby ecosystem but in other languages this is an antipattern.
It's much better experience to let the caller handle the error instead of crashing a thread.

Reply:
I mean it would be considered an SDK bug in that case — unhandled error that's crashing a thread. The best thing we can do is to generate a crash report and ask the user to send it our way. I don't expect the SDK user to know how to handle it, would you?